sys/include/irq_handler.h

```diff
@@ -0,0 +1,259 @@
+ * SPI or I2C. This type of modules also very often uses interrupts, which
+ * normally require access to the SPI or the I2C interfaces when an interrupt
+ * occurs, for example, to read status registers.
+ *
+ * Access to SPI and I2C interfaces is synchronized by mutual exclusion using
+ * mutexes. If one thread tries to access such an interface that is already
+ * being used by another thread, it will be blocked until the interface
+ * becomes available. Although this synchronization works in the thread
+ * context, it does not work in the interrupt context. Accessing such an
+ * interface within an interrupt service routine would interfere with an
+ * already existing interface access.
+ *
+ * The only solution to this problem is *not to call any function that
+ * interacts with a device directly from interrupt context*. Rather, the
+ * interrupt service routine only has to indicate the occurrence of the
+ * interrupt. The interrupt is then handled asynchronously in a thread.
+ *
+ * The problem is that many modules, such as driver modules, do not use their
+ * own thread, but run in the context of the calling thread. However, it can
+ * not be up to the application thread to handle the interrupts of such
+ * modules. The only solution would be for any module that uses interrupts
+ * with SPI or I2C interfaces to create its own interrupt handler thread.
+ * However, as the number of such modules increases, many resources (the
+ * thread context including the thread stack) are used only for interrupt
+ * handling.
+ *
+ * ## Solution
+ *
+ * The solution is to have only one thread with high priority for handling
+ * the interrupts of such modules. For this purpose, the module defines a
+ * prioritized interrupt event object for each interrupt source. The interrupt
+ * event object contains also a reference to the function by which it has to
+ * be handled.
+ *
+ * When an interrupt of the corresponding source occurs, the using module
+ * registers the interrupt event object associated with the interrupt source
+ * with the #irq_event_add function on the handler. The handler places the
+ * interrupt event object in a pending interrupt queue according to its
+ * priority. Each interrupt event object can be registered on the handler
+ * only once. That is, if the same interrupt occurs multiple times, only its
+ * first occurence is placed to the pending interrupt queue and is handled.
+ *
+ * When the interrupt handler thread gets the CPU, it processes all pending
+ * interrupt events in the order of their occurence before it yields.
+ *
+ * ## Usage
+ *
+ * This interrupt handler thread can be used not only for blocking functions
+ * such as SPI or I2C access, but also for other blocking functions. Please
+ * note, however, that the handling of interrupts is serialized by this
+ * module.
+ *
+ * To use the interrupt handler, the using module has to define a static
+ * interrupt event object of type #irq_event_t for each interrupt source. The
```

Comment:
I'd avoid the term "object", as this term is often understood as in "object oriented programming". Also "static" might be unclear, as the keyword `static` can either be used to control whether or not a symbol for linking is created (in front of global variables/functions), or if a local variable is allocated on the stack or in data/bss. Maybe something like "statically allocated structure" would be clearer?